sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala

```diff
@@ -737,108 +590,87 @@ class GpuOrcPartitionReader(
-   * when predicate push-down filters have been specified.
-   *
-   * @param evolution ORC SchemaEvolution
-   * @param useUTCTimestamp true if timestamps are UTC
-   * @return the search argument applier and search argument column mapping
-   */
-  private def getSearchApplier(
-      evolution: SchemaEvolution,
-      useUTCTimestamp: Boolean): (SargApplier, Array[Boolean]) = {
-    val searchArg = readerOpts.getSearchArgument
-    if (searchArg != null && orcReader.getRowIndexStride != 0) {
-      val sa = new SargApplier(searchArg, orcReader.getRowIndexStride, evolution,
-        orcReader.getWriterVersion, useUTCTimestamp)
-      // SargApplier.sargColumns is unfortunately not visible so we redundantly compute it here.
-      val filterCols = RecordReaderImpl.mapSargColumnsToOrcInternalColIdx(searchArg.getLeaves,
-        evolution)
-      val saCols = new Array[Boolean](evolution.getFileIncluded.length)
-      filterCols.foreach { i =>
-        if (i > 0) {
-          saCols(i) = true
-        }
-      }
-      (sa, saCols)
+  override def next(): Boolean = {
+    batch.foreach(_.close())
+    batch = None
+    if (ctx.blockIterator.hasNext) {
+      batch = readBatch()
     } else {
-      (null, null)
+      metrics(PEAK_DEVICE_MEMORY) += maxDeviceMemory
     }
+    // This is odd, but some operators return data even when there is no input so we need to
+    // be sure that we grab the GPU
+    GpuSemaphore.acquireIfNecessary(TaskContext.get())
+    batch.isDefined
   }
 
-  private def readPartFile(stripes: Seq[OrcOutputStripe]): (HostMemoryBuffer, Long) = {
-    withResource(new NvtxWithMetrics("Buffer file split", NvtxColor.YELLOW,
-        metrics("bufferTime"))) { _ =>
-      if (stripes.isEmpty) {
-        return (null, 0L)
-      }
+  override def close(): Unit = {
+    super.close()
+    ctx.orcReader.close()
+    ctx.dataReader.close()
+  }
 
-      val hostBufferSize = estimateOutputSize(stripes)
-      var succeeded = false
-      val hmb = HostMemoryBuffer.allocate(hostBufferSize)
-      try {
-        val out = new HostMemoryOutputStream(hmb)
-        writeOrcOutputFile(out, stripes)
-        succeeded = true
-        (hmb, out.getPos)
-      } finally {
-        if (!succeeded) {
-          hmb.close()
+  private def readBatch(): Option[ColumnarBatch] = {
+    withResource(new NvtxRange("ORC readBatch", NvtxColor.GREEN)) { _ =>
+      val currentStripes = populateCurrentBlockChunk(ctx.blockIterator, maxReadBatchSizeRows,
+        maxReadBatchSizeBytes)
+      if (readDataSchema.isEmpty) {
+        // not reading any data, so return a degenerate ColumnarBatch with the row count
+        val numRows = currentStripes.map(_.infoBuilder.getNumberOfRows).sum
+        Some(new ColumnarBatch(Array.empty, numRows.toInt))
+      } else {
+        val table = readToTable(currentStripes)
+        try {
+          table.map(GpuColumnVector.from(_, readDataSchema.toArray.map(_.dataType)))
```

Comment:
NIT:
There is already an API `extractTypes` to convert the schema to array of DataType.
```suggestion
          table.map(GpuColumnVector.from(_, extractTypes(readDataSchema)))
```